crates/core/src/rpc/response.rs

```diff
@@ -1,37 +1,244 @@
+	pub execution_time: Option<Duration>,
+}
+
+impl DbResultStats {
+	pub fn with_execution_time(mut self, execution_time: Duration) -> Self {
+		self.execution_time = Some(execution_time);
+		self
+	}
+}
 
 /// The data returned by the database
 // The variants here should be in exactly the same order as `crate::engine::remote::ws::Data`
 // In future, they will possibly be merged to avoid having to keep them in sync.
 #[revisioned(revision = 1)]
-#[derive(Debug, Serialize)]
-pub enum Data {
+#[derive(Debug, Serialize, Deserialize)]
+pub enum DbResult {
 	/// Generally methods return a `expr::Value`
 	Other(Value),
 	/// The query methods, `query` and `query_with` return a `Vec` of responses
-	Query(Vec<dbs::Response>),
+	Query(Vec<dbs::QueryResult>),
 	/// Live queries return a notification
-	Live(Notification),
+	Live(PublicNotification),
 	// Add new variants here
 }
 
-impl Data {
-	pub fn into_value(self) -> Value {
+impl SurrealValue for DbResult {
+	fn kind_of() -> Kind {
+		Kind::Any
+	}
+
+	fn is_value(_value: &Value) -> bool {
+		true
+	}
+
+	fn into_value(self) -> Value {
 		match self {
-			Data::Query(v) => v.into_iter().map(|x| x.into_value()).collect(),
-			Data::Live(v) => Value::from(Object(map! {
-				"id".to_owned() => v.id.into(),
-				"action".to_owned() => v.action.to_string().into(),
+			DbResult::Query(v) => {
+				let converted: Vec<Value> = v.into_iter().map(|x| x.into_value()).collect();
+				Value::Array(surrealdb_types::Array::from_values(converted))
+			}
+			DbResult::Live(v) => Value::from(surrealdb_types::Object::from_map(map! {
+				"id".to_owned() => Value::Uuid(surrealdb_types::Uuid(v.id.0)),
+				"action".to_owned() => Value::String(v.action.to_string()),
 				"record".to_owned() => v.record,
 				"result".to_owned() => v.result,
 
 			})),
-			Data::Other(v) => v,
+			DbResult::Other(v) => v,
+		}
+	}
+
+	fn from_value(value: Value) -> anyhow::Result<Self> {
+		match value {
+			Value::Array(arr) => {
+				let results = arr
+					.inner()
+					.iter()
+					.cloned()
+					.map(QueryResult::from_value)
+					.collect::<anyhow::Result<Vec<_>>>()?;
+				Ok(DbResult::Query(results))
+			}
+			Value::Object(obj) => {
+				// Check if this is a Live result
+				if obj.get("id").is_some() && obj.get("action").is_some() {
+					let mut obj = obj.inner().clone();
+					let id = obj.remove("id").context("Missing id")?;
+					let action = obj.remove("action").context("Missing action")?;
+					let record = obj.remove("record").unwrap_or(Value::None);
+					let result = obj.remove("result").unwrap_or(Value::None);
+
+					let Value::Uuid(uuid) = id else {
+						anyhow::bail!("Expected UUID for id field");
+					};
+					let Value::String(action_str) = action else {
+						anyhow::bail!("Expected string for action field");
+					};
+
+					// Parse action string to PublicAction
+					let action = match action_str.as_str() {
+						"CREATE" => crate::types::PublicAction::Create,
```

Comment:
perhaps PublicAction can implement SurrealValue too

Reply:
It does now.